pairing_api/src/pairing.rs

```diff
@@ -0,0 +1,456 @@
+                        code: 6000,
+                        message: "User requested disconnect".to_owned(),
+                    }),
+                    client,
+                )
+                .await?;
+            };
+        }
+
+        {
+            client
+                .unsubscribe(topic.into())
+                .await
+                .map_err(PairingClientError::SubscriptionError)?;
+        };
+
+        Ok(())
+    }
+
+    /// Used to update the lifetime of a pairing.
+    /// https://specs.walletconnect.com/2.0/specs/clients/core/pairing/rpc-methods#wc_pairingextend
+    pub async fn extend(
+        &self,
+        topic: &str,
+        expiry: u64,
+        client: &Client,
+    ) -> Result<(), PairingClientError> {
+        let extend_request = RequestParams::PairingExtend(PairingExtendRequest { expiry });
+        self.publish_request(topic, extend_request, client).await?;
+
+        Ok(())
+    }
+
+    /// Private function to publish a request.
+    async fn publish_request(
+        &self,
+        topic: &str,
+        params: RequestParams,
+        client: &Client,
+    ) -> Result<(), PairingClientError> {
+        let irn_metadata = params.irn_metadata();
+        let message_id = MessageIdGenerator::new().next();
+        let request = Request::new(message_id, params.into());
+        // Publish the encrypted message
+        self.publish_payload(topic, irn_metadata, Payload::Request(request), client)
+            .await?;
+
+        Ok(())
+    }
+
+    /// Private function to publish a request response.
+    pub async fn publish_response(
+        &self,
+        topic: &str,
+        params: ResponseParamsSuccess,
+        message_id: MessageId,
+        client: &Client,
+    ) -> Result<(), PairingClientError> {
+        let irn_metadata = params.irn_metadata();
+        let response = Response::Success(SuccessfulResponse {
+            id: message_id,
+            jsonrpc: JSON_RPC_VERSION_STR.into(),
+            result: serde_json::to_value(params)
+                .map_err(|err| PairingClientError::EncodeError(err.to_string()))?,
+        });
+
+        // Publish the encrypted message
+        self.publish_payload(topic, irn_metadata, Payload::Response(response), client)
+            .await?;
+
+        Ok(())
+    }
+
+    /// Private function to publish a payload.
+    async fn publish_payload(
+        &self,
+        topic: &str,
+        irn_metadata: IrnMetadata,
+        payload: Payload,
+        client: &Client,
+    ) -> Result<(), PairingClientError> {
+        // try to extend session before updating local store.
+        let sym_key = {
+            let pairings = self.pairings.lock().await;
+            let pairing = pairings
+                .get(topic)
+                .ok_or_else(|| PairingClientError::PairingNotFound)?;
+            hex::decode(pairing.sym_key.clone()).map_err(|err| {
+                PairingClientError::EncodeError(format!("Failed to decode sym_key: {:?}", err))
```

Comment:
The name is EncodeError but the description says it's 'Failed to decode'.
Maybe good to add a dedicated error DecodeSymKeyError

Reply:
iterated and fixed earlier thanks